### PR TITLE
Allow multiple SubmapTextures.

### PR DESCRIPTION
**Pull request by gaschler**
This changes submap_visualization.proto that multiple textures can be provided.
As of now, the first texture is generated from the high resolution grid,
which is the same as before.
The second texture is generated from the low resolution grid.

### CODE REVIEW

File: cartographer/mapping_3d/submaps.cc [hunk 1]
```diff
@@ -146,6 +146,39 @@ string ComputePixelValues(
   return cell_data;
 }
 
+void AddTextureToTextureProto(const HybridGrid& grid,
```
Comment (wohe) [inline cartographer/mapping_3d/submaps.cc]: Run `clang-format`? `hybrid_grid`?

File: cartographer/mapping/proto/submap_visualization.proto [hunk 2]
```diff
@@ -43,20 +43,25 @@ message SubmapQuery {
     // Version of the given submap, higher means newer.
     optional int32 submap_version = 2;
 
-    // GZipped map data, in row-major order, starting with (0,0). Each cell
-    // consists of two bytes: value (premultiplied by alpha) and alpha.
-    optional bytes cells = 3;
+    message SubmapTexture {
+      // GZipped map data, in row-major order, starting with (0,0). Each cell
+      // consists of two bytes: value (premultiplied by alpha) and alpha.
+      optional bytes cells = 3;
```
[the remark this answers is not in the document]
Comment (gaschler) [reply cartographer/mapping/proto/submap_visualization.proto]: Done

File: cartographer/mapping/proto/submap_visualization.proto [hunk 3]
```diff
@@ -43,20 +43,25 @@ message SubmapQuery {
     // Version of the given submap, higher means newer.
     optional int32 submap_version = 2;
 
-    // GZipped map data, in row-major order, starting with (0,0). Each cell
-    // consists of two bytes: value (premultiplied by alpha) and alpha.
-    optional bytes cells = 3;
+    message SubmapTexture {
```
[the remark this answers is not in the document]
Comment (gaschler) [reply cartographer/mapping/proto/submap_visualization.proto]: Done

File: cartographer/mapping_2d/submaps.cc [hunk 4]
```diff
@@ -113,17 +113,18 @@ void Submap::ToResponseProto(
       cells.push_back(0);                                    // alpha
     }
   }
-  common::FastGzipString(cells, response->mutable_cells());
+  mapping::proto::SubmapQuery::Response::SubmapTexture* texture = response->add_textures();
```
[the remark this answers is not in the document]
Comment (gaschler) [reply cartographer/mapping_2d/submaps.cc]: Done

File: cartographer/mapping_3d/submaps.cc [hunk 5]
```diff
@@ -194,34 +227,11 @@ void Submap::ToResponseProto(
...
-      width, height, min_index, max_index, voxel_indices_and_probabilities);
-  const string cell_data = ComputePixelValues(accumulated_pixel_data);
-
-  common::FastGzipString(cell_data, response->mutable_cells());
-  *response->mutable_slice_pose() = transform::ToProto(
-      global_submap_pose.inverse() *
-      transform::Rigid3d::Translation(Eigen::Vector3d(
-          max_index.x() * resolution, max_index.y() * resolution,
-          global_submap_pose.translation().z())));
+  AddTextureToTextureProto(high_resolution_hybrid_grid_,
```
[the remark this answers is not in the document]
Comment (gaschler) [reply cartographer/mapping_3d/submaps.cc]: Done

File: cartographer/mapping_3d/submaps.cc (hunk 1 above)
Comment (gaschler) [reply cartographer/mapping_3d/submaps.cc]: Done

**Comment by gaschler** [comment]
Thanks for the corrections, PTAL at the changes.

**Pull request merged**